File: examples/playground/src/MyStateBag.ts
import {Color, File, FontRegistry, Haiku, Image, IOSFonts, Lottie, Palette, SVG, TextStyle} from '@diez/designsystem';
import {Component, expression, Integer, method, property, shared} from '@diez/engine';
import {easeInOutExpo} from 'just-curves';

class MyPalette extends Component<Palette> {
  @shared helloRValue!: number;
  @property hello = expression<Color>((helloRValue: number) => Color.rgba(helloRValue, 0, 0, 1));
}

export class MyStateBag extends Component {
  @property palette = new MyPalette();

  @property helloRValue = 255;

  private name = 'Diez';

  @property text = `Hello ${this.name}`;

  @property numbers: Integer[] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

  // Try changing this to `assets/images/diez`!
  @property image = Image.responsive('assets/images/haiku', 'jpg');

  @property svg = new SVG({src: 'assets/images/rat.svg'});

  @property lottie = new Lottie({
    file: new File({src: 'assets/lottie/loading-pizza.json'}),
  });

  // Makes Android fonts available on iOS.
  @property fontRegistry = FontRegistry.fromFiles(
    'assets/fonts/Roboto-Black.ttf',
    'assets/fonts/Roboto-BlackItalic.ttf',
    'assets/fonts/Roboto-Bold.ttf',
    'assets/fonts/Roboto-BoldItalic.ttf',
    'assets/fonts/Roboto-Italic.ttf',
    'assets/fonts/Roboto-Light.ttf',
    'assets/fonts/Roboto-LightItalic.ttf',
    'assets/fonts/Roboto-Medium.ttf',
    'assets/fonts/Roboto-MediumItalic.ttf',
    'assets/fonts/Roboto-Regular.ttf',
    'assets/fonts/Roboto-Thin.ttf',
    'assets/fonts/Roboto-ThinItalic.ttf',
  );

  @property textStyle = new TextStyle({
    fontName: IOSFonts.Helvetica,
    fontSize: 50,
    color: this.palette.hello,
  });

  @property haiku = new Haiku({
    component: '@haiku/taylor-testthang',
  });

  @method async tap () {
    await this.palette.hello.tween(
      {h: 0.3},
      // Default curve is linear!
      {duration: 5000},
    );
    await this.palette.hello.tween(
      // You can tween multiple values at once!
      {h: 0.9, l: 0.5},
      {
        duration: 5000,
        // You can specify any pure function (number) -> number as a curve!
        curve: easeInOutExpo,
      },
    );
    await this.palette.tween(
      // You can tween an entire component instead of its inner state!
      {hello: Color.rgba(255, 0, 0, 1)},
      {duration: 5000},
    );
  }
}

File: examples/playground/android/diez/src/main/java/org/diez/MyStateBag.kt
package org.diez

import com.squareup.moshi.Moshi
import android.graphics.Color as CoreColor

data class MyPalette(@QualifiedColor var hello: Color = ColorAdapter.hsla(
    WireColor(
        0F,
        1F,
        0.5F,
        1F
    )
))

data class MyStateBag(
    var palette: MyPalette = MyPalette(),
    var copy: String = "Hello Diez",
    var image: Image = Image(File("assets/images/haiku.jpg"), 246, 246, 3F),
    var fontRegistry: FontRegistry = FontRegistry(
        arrayOf<File>(
            File("assets/fonts/Roboto-Black.ttf"),
            File("assets/fonts/Roboto-BlackItalic.ttf"),
            File("assets/fonts/Roboto-Bold.ttf"),
            File("assets/fonts/Roboto-BoldItalic.ttf"),
            File("assets/fonts/Roboto-Italic.ttf"),
            File("assets/fonts/Roboto-Light.ttf"),
            File("assets/fonts/Roboto-LightItalic.ttf"),
            File("assets/fonts/Roboto-Medium.ttf"),
            File("assets/fonts/Roboto-MediumItalic.ttf"),
            File("assets/fonts/Roboto-Regular.ttf"),
            File("assets/fonts/Roboto-Thin.ttf"),
            File("assets/fonts/Roboto-ThinItalic.ttf")
        )
    ),
    var textStyle: TextStyle = TextStyle(
        "Helvetica",
        50F,
        ColorAdapter.hsla(WireColor(0F, 1F, 0.5F, 1F))
    ),
    var haiku: Haiku = Haiku("@haiku/taylor-testthang"),
    var svg: SVG = SVG("assets/images/rat.svg"),
    var lottie: Lottie = Lottie(File("assets/lottie/loading-pizza.json"))
) : StateBag {
    override fun registerAdapters(builder: Moshi.Builder) {
        builder.add(colorAdapter)
    }

    override val name = "MyStateBag"
}
